File: demos/04-arrays-slices/arrays_slices.ts
// 与 arrays_slices.go 逐段对照

function demoArrayReference(): void {
  console.log("---------demoArrayReference---------");
  const languages: [string, string, string] = ["TypeScript", "Go", "Rust"];
  const alias = languages;
  alias[0] = "JavaScript";
  console.log("original:", languages);
  console.log("alias:   ", alias);
}

function demoArrayCopy(): void {
  console.log("---------demoArrayCopy---------");
  const languages = ["TypeScript", "Go", "Rust"];
  const copied = [...languages];
  copied[0] = "JavaScript";
  console.log("original:", languages);
  console.log("copied:  ", copied);
}

function demoSliceCopy(): void {
  console.log("---------demoSliceCopy---------");
  const languages = ["TypeScript", "Go", "Rust"];
  const rangeCopy = languages.slice(1, 3);
  rangeCopy[0] = "Golang";
  console.log("array:     ", languages);
  console.log("range copy:", rangeCopy);
}

function demoPush(): void {
  console.log("---------demoPush---------");
  const numbers: number[] = [];
  const newLength = numbers.push(42);
  console.log("length:", newLength);
  console.log("array: ", numbers);
}

demoArrayReference();
demoArrayCopy();
demoSliceCopy();
demoPush();

File: demos/04-arrays-slices/arrays_slices.go
package main

import "fmt"

func main() {
	demoArrayCopy()
	demoSliceView()
	demoNilSlice()
	demoAppendCapacity()
	demoIndependentCopy()
}

func demoArrayCopy() {
	fmt.Println("---------demoArrayCopy---------")
	languages := [3]string{"TypeScript", "Go", "Rust"}
	copied := languages
	copied[0] = "JavaScript"
	fmt.Println("original:", languages)
	fmt.Println("copied:  ", copied)
}

func demoSliceView() {
	fmt.Println("---------demoSliceView---------")
	languages := []string{"TypeScript", "Go", "Rust"}
	view := languages[1:3]
	view[0] = "Golang"
	fmt.Println("slice:", languages)
	fmt.Println("view: ", view)
}

func demoNilSlice() {
	fmt.Println("---------demoNilSlice---------")
	var numbers []int
	fmt.Println("before append:", numbers == nil, len(numbers), cap(numbers))
	numbers = append(numbers, 42)
	fmt.Println("after append: ", numbers)
}

func demoAppendCapacity() {
	fmt.Println("---------demoAppendCapacity---------")
	base := make([]string, 2, 3)
	base[0] = "TypeScript"
	base[1] = "Go"
	sharedView := base[:3]

	grown := append(base, "Rust")
	grown[0] = "JavaScript"
	fmt.Println("within capacity:", sharedView)

	moved := append(grown, "Python")
	moved[0] = "TS"
	fmt.Println("after growth:   ", moved)
	fmt.Println("old storage:    ", sharedView)
}

func demoIndependentCopy() {
	fmt.Println("---------demoIndependentCopy---------")
	original := []int{1, 2, 3}
	cloned := make([]int, len(original))
	copied := copy(cloned, original)
	cloned[0] = 99
	fmt.Println("copied elements:", copied)
	fmt.Println("original:", original)
	fmt.Println("cloned:  ", cloned)
}
